File: _submissions/18UEE071_Ayush/Day7/removeHalfNodes.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
TreeNode* Solution::solve(TreeNode* root) 
{
    if(!root) return NULL;
    
    if(!root->left and !root->right)
        return root;
    

    root->left = solve(root->left);
    root->right = solve(root->right);

    if(!root->left or !root->right)
    {
        if(!root->left)
            return root->right;
        if(!root->right)
            return root->left;
    }
    return root;
}

File: _submissions/19UCS159_Tharun/Day7/removeHalfNodes.java

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
    left = null;
    right = null;
  }
}

public class removeHalfNodes {
  public TreeNode solve(TreeNode A) {
    // leaf
    if (A.left == null && A.right == null) {
      return A;
    }
    if (A.left == null)
      return solve(A.right);
    if (A.right == null)
      return solve(A.left);

    A.left = solve(A.left);
    A.right = solve(A.right);
    return A;
  }
}
